Switch TORCH_TRACE to accept a directory by default

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121331

Directory is better because it works smoothly with distributed
runs; otherwise you'd need to modify torchrun to setup distinct
log names for each file.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

Differential Revision: [D54597814](https://our.internmc.facebook.com/intern/diff/D54597814)

ezyang committed:
Switch TORCH_TRACE to accept a directory by default

Directory is better because it works smoothly with distributed
runs; otherwise you'd need to modify torchrun to setup distinct
log names for each file.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54597814).

Skylion007 commented on torch/_logging/_internal.py:
```diff
@@ -961,46 +955,46 @@ def close(self):
 
-            if ok:
+            if self.root_dir is not None:
+                if not os.path.exists(self.root_dir):
+                    os.makedirs(self.root_dir, exist_ok=True)
```
```suggestion
                os.makedirs(self.root_dir, exist_ok=True)
```

albanD reviewed (approved):
SGTM
What is the BC-story for this?

ezyang committed:
Update on "Switch TORCH_TRACE to accept a directory by default"


Directory is better because it works smoothly with distributed
runs; otherwise you'd need to modify torchrun to setup distinct
log names for each file.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

Differential Revision: [D54597814](https://our.internmc.facebook.com/intern/diff/D54597814)

[ghstack-poisoned]

ezyang commented:
@albanD the envvar has existed for a week and hasn't been documented anywhere, so I claim special privilege!!

albanD commented:
Sounds good!

ezyang commented:
@pytorchbot merge

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54597814).

ezyang commented:
@pytorchbot merge
